low-level/sensors/impl/centos_7/service_monitor.py

```diff
@@ -0,0 +1,411 @@
+        self._bus = SystemBus()
+        systemd = self._bus.get_object("org.freedesktop.systemd1", "/org/freedesktop/systemd1")
+        self._manager = Interface(systemd, dbus_interface='org.freedesktop.systemd1.Manager')
+
+        self.remove_disabled_services()
+
+        return True
+
+    def remove_disabled_services(self):
+        """Remove `disabled` services from the list of services to monitor."""
+        temp = self.services_to_monitor.copy()
+        for service in temp:
+            try:
+                if 'disabled' in str(self._manager.GetUnitFileState(service)):
+                    self.services_to_monitor.remove(service)
+            except DBusException:
+                self.services_to_monitor.remove(service)
+
+    def run(self):
+        logger.info(f"Monitoring Services : {self.services_to_monitor}")
+        try:
+            # Register all the services to signal of 'PropertiesChanged' and
+            # raise an alert if some service is not active on initially or if
+            # Unit is not found for the service
+            temp_ser_to_mon = self.services_to_monitor.copy()
+            for service in temp_ser_to_mon:
+                err = self.connect_to_prop_changed_signal(service)
+                if err:
+                    self.raise_alert(service, "N/A", "N/A", "N/A", "N/A",
+                        "N/A", "N/A", 0)
+                    logger.error(f"{service} is not active initially. \n Error {err}")
+
+            logger.debug(f"failed_services : {self.failed_services}")
+            logger.debug(f"services_to_monitor : {self.services_to_monitor}")
+
+            # Retrieve the main loop which will be called in the run method
+            self._loop = gobject.MainLoop()
+
+            # Initialize the gobject threads and get its context
+            gobject.threads_init()
+            context = self._loop.get_context()
+
+            delay = 0
+            while self.is_running():
+                # At interval of 'thread_sleep' check for events occured for
+                # registered services and process them(call on_pro_changed())
+                context.iteration(False)
+                time.sleep(self.thread_sleep)
+
+                # At interval of 'polling_freqency' process unregistered
+                # services and services with not-active state. 
+                if delay == self.polling_frequency:
+                    delay = 0
+
+                    temp_ser_to_mon = self.services_to_monitor.copy()
+                    for service in temp_ser_to_mon:
+                        self.connect_to_prop_changed_signal(service)
+
+                    self.check_notactive_services()
+
+                delay+=1
+
+            logger.debug("ServiceMonitor gracefully breaking out " +\
+                                "of dbus Loop, not restarting.")
+        except Exception as e:
+            if self.is_running:
+                logger.debug("Ungracefully breaking out of " +\
```

Comment:
Can we have specific exception handling, based on possible exceptions, this one guideline came from Ujjwal last week.

Also, this should be logger.error as error should be known for the failure, dont keep such imp messages for debug alone.

Reply:
Sure..